pc/randomsample.go

```diff
@@ -0,0 +1,64 @@
+package pc
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/seqsense/pcgol/mat"
+)
+
+// NewVec3RandomSampleIterator creates Vec3 iterator with random sampling.
+// Sampled data won't duplicate.
+func NewVec3RandomSampleIterator(ra Vec3RandomAccessor, ratio float32) Vec3ConstForwardIterator {
+	switch {
+	case ratio <= 0:
+		return &vec3EmptyIterator{}
+	case ratio >= 1:
+		if it, ok := ra.(Vec3ConstForwardIterator); ok {
+			return it
+		}
+		return &vec3RandomAccessToIterator{ra: ra}
+	}
+	expectedInterval := 1 / ratio
+	return &vec3RandomSampleIterator{
+		ra:     ra,
+		Lambda: 1 / (expectedInterval - 1),
+	}
+}
+
+type vec3RandomSampleIterator struct {
+	ra Vec3RandomAccessor
+
+	Lambda float32
+	pos    float32
+}
+
+func (s *vec3RandomSampleIterator) Incr() {
+	// Jump position by A-ExpJ.
+	inc := -float32(math.Log(float64(1-rand.Float32()))) / s.Lambda
```

Comment:
I'm not familiar with the algorithm but why `1-rand.Float32()` instead of `rand.Float32()` which seems to be used in https://utopia.duth.gr/~pefraimi/research/data/2007EncOfAlg.pdf.

Reply:
`rand.Float32()` returns 0.0 <= v < 1.0, but `Log(0)` is undefined.

Reply:
If using `rand.Float32()` directly, NaN check should be added.